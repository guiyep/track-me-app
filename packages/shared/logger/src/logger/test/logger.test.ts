import {
  global,
  warn,
  log,
  error,
  asyncFunc,
  syncFunc,
  loggerDecorator,
} from '../index';

const getSnapshotName = (testName: string): string => {
  const folderName = 'logger';
  return `${folderName}/${testName}`;
};

describe('Logger', () => {
  let consoleLogSpy: jest.SpyInstance;
  let consoleWarnSpy: jest.SpyInstance;
  let consoleErrorSpy: jest.SpyInstance;

  beforeEach(() => {
    consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
    consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
  });

  afterEach(() => {
    jest.clearAllMocks();
    global.setLoggerStatus(true);
  });

  describe('logger status', () => {
    it('should not log when logger is inactive', () => {
      global.setLoggerStatus(false);
      log({ message: 'test message' });
      expect(consoleLogSpy).not.toHaveBeenCalled();
    });

    it('should log when logger is active', () => {
      global.setLoggerStatus(true);
      log({ message: 'test message' });
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-status-active'),
      );
    });
  });

  describe('log function', () => {
    it('should log message without object', () => {
      log({ message: 'test message' });
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('log-without-object'),
      );
    });

    it('should log message with object', () => {
      const testObj = { key: 'value' };
      log({ message: 'test message' }, testObj);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('log-with-object'),
      );
    });

    it('should log message with primitive value', () => {
      log({ message: 'test message' }, 123);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('log-with-primitive'),
      );
    });

    it('should log message with name prefix', () => {
      log({ message: 'test message' }, undefined, 'TestModule');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('log-with-prefix'),
      );
    });

    it('should handle null object', () => {
      log({ message: 'test message' }, null);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('log-with-null'),
      );
    });
  });

  describe('warn function', () => {
    it('should warn message without object', () => {
      warn({ message: 'test warning' });
      expect(consoleWarnSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('warn-without-object'),
      );
    });

    it('should warn message with object', () => {
      const testObj = { key: 'value' };
      warn({ message: 'test warning' }, testObj);
      expect(consoleWarnSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('warn-with-object'),
      );
    });

    it('should warn message with name prefix', () => {
      warn({ message: 'test warning' }, undefined, 'TestModule');
      expect(consoleWarnSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('warn-with-prefix'),
      );
    });
  });

  describe('error function', () => {
    it('should error message without object', () => {
      error({ message: 'test error' });
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('error-without-object'),
      );
    });

    it('should error message with error object', () => {
      const testError = new Error('test error');
      error({ message: 'test error' }, testError);
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('error-with-error-object'),
      );
    });

    it('should error message with name prefix', () => {
      error({ message: 'test error' }, undefined, 'TestModule');
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('error-with-prefix'),
      );
    });
  });

  describe('asyncFunc wrapper', () => {
    it('should log function execution start and completion', async () => {
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = asyncFunc(testFunction);

      const result = await wrappedFunction('test');

      expect(result).toBe('TEST');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('async-func-success'),
      );
    });

    it('should log error when function fails', async () => {
      const testError = new Error('Test error');
      const failingFunction = () => {
        throw testError;
      };
      const wrappedFunction = asyncFunc(failingFunction);

      await expect(wrappedFunction()).rejects.toThrow(testError);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('async-func-error-log'),
      );
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('async-func-error-error'),
      );
    });

    it('should use custom name when provided', async () => {
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = asyncFunc(testFunction, 'CustomName');

      await wrappedFunction('test');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('async-func-custom-name'),
      );
    });
  });

  describe('syncFunc wrapper', () => {
    it('should log function execution start and completion', () => {
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = syncFunc(testFunction);

      const result = wrappedFunction('test');

      expect(result).toBe('TEST');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('sync-func-success'),
      );
    });

    it('should log error when function fails', () => {
      const testError = new Error('Test error');
      const failingFunction = () => {
        throw testError;
      };
      const wrappedFunction = syncFunc(failingFunction);

      expect(() => wrappedFunction()).toThrow(testError);
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('sync-func-error-log'),
      );
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('sync-func-error-error'),
      );
    });

    it('should use custom name when provided', () => {
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = syncFunc(testFunction, 'CustomName');

      wrappedFunction('test');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('sync-func-custom-name'),
      );
    });
  });

  describe('loggerDecorator', () => {
    it('should create logger with custom name', () => {
      const customLogger = loggerDecorator('TestModule');

      customLogger.log({ message: 'test message' });
      customLogger.warn({ message: 'test warning' });
      customLogger.error({ message: 'test error' });

      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-decorator-log'),
      );
      expect(consoleWarnSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-decorator-warn'),
      );
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-decorator-error'),
      );
    });

    it('should wrap async function with custom name', async () => {
      const customLogger = loggerDecorator('TestModule');
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = customLogger.asyncFunc(testFunction);

      await wrappedFunction('test');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-decorator-async'),
      );
    });

    it('should wrap sync function with custom name', () => {
      const customLogger = loggerDecorator('TestModule');
      const testFunction = (param: string) => param.toUpperCase();
      const wrappedFunction = customLogger.syncFunc(testFunction);

      wrappedFunction('test');
      expect(consoleLogSpy.mock.calls).toMatchSnapshot(
        getSnapshotName('logger-decorator-sync'),
      );
    });
  });
});
