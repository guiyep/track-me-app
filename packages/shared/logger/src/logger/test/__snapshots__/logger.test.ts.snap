// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger asyncFunc wrapper should log error when function fails: logger/async-func-error-error 1`] = `
[
  [
    ""failingFunction"/async_error",
  ],
  [
    ""failingFunction"/async_error",
    [Error: Test error],
  ],
  [
    ""failingFunction"/async_error",
    undefined,
  ],
]
`;

exports[`Logger asyncFunc wrapper should log error when function fails: logger/async-func-error-log 1`] = `
[
  [
    ""failingFunction"/async_start",
    "{}",
  ],
]
`;

exports[`Logger asyncFunc wrapper should log function execution start and completion: logger/async-func-success 1`] = `
[
  [
    ""testFunction"/async_start",
    "{"0":"test"}",
  ],
  [
    ""testFunction"/async_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger asyncFunc wrapper should use custom name when provided: logger/async-func-custom-name 1`] = `
[
  [
    ""CustomName"/async_start",
    "{"0":"test"}",
  ],
  [
    ""CustomName"/async_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger error function should error message with error object: logger/error-with-error-object 1`] = `
[
  [
    "test error",
  ],
  [
    "test error",
    [Error: test error],
  ],
  [
    "test error",
    undefined,
  ],
]
`;

exports[`Logger error function should error message with name prefix: logger/error-with-prefix 1`] = `
[
  [
    "TestModule/error test error",
  ],
  [
    "TestModule/error test error",
    undefined,
  ],
]
`;

exports[`Logger error function should error message without object: logger/error-without-object 1`] = `
[
  [
    "test error",
  ],
  [
    "test error",
    undefined,
  ],
]
`;

exports[`Logger log function should handle null object: logger/log-with-null 1`] = `
[
  [
    "test message",
  ],
  [
    "test message",
    "null",
  ],
]
`;

exports[`Logger log function should log message with name prefix: logger/log-with-prefix 1`] = `
[
  [
    "TestModule/log test message",
  ],
  [
    "TestModule/log test message",
    undefined,
  ],
]
`;

exports[`Logger log function should log message with object: logger/log-with-object 1`] = `
[
  [
    "test message",
    "{"key":"value"}",
  ],
]
`;

exports[`Logger log function should log message with primitive value: logger/log-with-primitive 1`] = `
[
  [
    "test message",
    123,
  ],
]
`;

exports[`Logger log function should log message without object: logger/log-without-object 1`] = `
[
  [
    "test message",
  ],
  [
    "test message",
    undefined,
  ],
]
`;

exports[`Logger logger status should log when logger is active: logger/logger-status-active 1`] = `
[
  [
    "test message",
  ],
  [
    "test message",
    undefined,
  ],
]
`;

exports[`Logger loggerDecorator should create logger with custom name: logger/logger-decorator-error 1`] = `
[
  [
    "TestModule/error test error",
  ],
  [
    "TestModule/error test error",
    undefined,
  ],
]
`;

exports[`Logger loggerDecorator should create logger with custom name: logger/logger-decorator-log 1`] = `
[
  [
    "TestModule/log test message",
  ],
  [
    "TestModule/log test message",
    undefined,
  ],
]
`;

exports[`Logger loggerDecorator should create logger with custom name: logger/logger-decorator-warn 1`] = `
[
  [
    "TestModule/warn test warning",
  ],
  [
    "TestModule/warn test warning",
    undefined,
  ],
]
`;

exports[`Logger loggerDecorator should wrap async function with custom name: logger/logger-decorator-async 1`] = `
[
  [
    ""TestModule"/async_start",
    "{"0":"test"}",
  ],
  [
    ""TestModule"/async_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger loggerDecorator should wrap sync function with custom name: logger/logger-decorator-sync 1`] = `
[
  [
    ""TestModule"/sync_start",
    "["test"]",
  ],
  [
    ""TestModule"/sync_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger syncFunc wrapper should log error when function fails: logger/sync-func-error-error 1`] = `
[
  [
    ""failingFunction"/sync_error",
  ],
  [
    ""failingFunction"/sync_error",
    [Error: Test error],
  ],
  [
    ""failingFunction"/sync_error",
    undefined,
  ],
]
`;

exports[`Logger syncFunc wrapper should log error when function fails: logger/sync-func-error-log 1`] = `
[
  [
    ""failingFunction"/sync_start",
    "[]",
  ],
]
`;

exports[`Logger syncFunc wrapper should log function execution start and completion: logger/sync-func-success 1`] = `
[
  [
    ""testFunction"/sync_start",
    "["test"]",
  ],
  [
    ""testFunction"/sync_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger syncFunc wrapper should use custom name when provided: logger/sync-func-custom-name 1`] = `
[
  [
    ""CustomName"/sync_start",
    "["test"]",
  ],
  [
    ""CustomName"/sync_end",
    "{"result":"TEST","args":["test"]}",
  ],
]
`;

exports[`Logger warn function should warn message with name prefix: logger/warn-with-prefix 1`] = `
[
  [
    "TestModule/warn test warning",
  ],
  [
    "TestModule/warn test warning",
    undefined,
  ],
]
`;

exports[`Logger warn function should warn message with object: logger/warn-with-object 1`] = `
[
  [
    "test warning",
    "{"key":"value"}",
  ],
]
`;

exports[`Logger warn function should warn message without object: logger/warn-without-object 1`] = `
[
  [
    "test warning",
  ],
  [
    "test warning",
    undefined,
  ],
]
`;
